texk/ptexenc/ptekf.1
.TH PTEKF 1
\"=====================================================================
.if t .ds TX \fRT\\h'-0.1667m'\\v'0.20v'E\\v'-0.20v'\\h'-0.125m'X\fP
.if n .ds TX TeX
\"=====================================================================
.SH NAME
ptekf \- p\*(TX-enc Kanji Filter
.SH SYNOPSIS
.HP
\fBptekf\fR [\fIoptions\fR] [\fB\-\-\fR] \fIin_file1\fR [\fIin_file2 ...\fR]
.HP
\fBptekf\fR \-\-help | \-\-version
.SH DESCRIPTION
\fBptekf\fR is a simple character encoding converter for input/output files of p\*(TX-family programs.
This utility can be used to check procedures of character encoding conversion and guessing encoding \
which are impremented in p\*(TX-family i.e. \fB(e)(u)ptex\fR, \fB(u)pbibtex\fR, \fBmendex\fR and so on.
.SH OPTIONS
.PP
.TP
\fB\-j\fR
Specify output encoding ISO-2022-JP.
.TP
\fB\-s\fR
Specify output encoding Shift_JIS.
.TP
\fB\-e\fR
Specify output encoding EUC-JP.
.TP
\fB\-u\fR
Specify output encoding UTF-8.
.TP
\fB\-J\fR
Specify input encoding ISO-2022-JP.
.TP
\fB\-S\fR
Specify input encoding Shift_JIS.
.TP
\fB\-E\fR
Specify input encoding EUC-JP.
.TP
\fB\-U\fR
Specify input encoding UTF-8.
.TP
\fB\-g\fR  \fB\-\-guess\fR
Guess the input encoding (no conversion).
.TP
\fB\-G\fR
Guess the input encoding and output to stdout or files.
.TP
\fB\-b\fR  \fB\-\-buffer\fR
Output internal buffer without code conversion.
.TP
\fB\-v\fR  \fB\-\-version\fR
Print version number.
.TP
\fB\-h\fR  \fB\-\-help\fR
Print help message.
.TP
Default input/output encoding depends on kpathsearch parameters \fBPTEX_KANJI_ENC\fR and \fBguess_input_kanji_encoding\fR.
.SH SEE ALSO
eptex(1), euptex(1), pbibtex(1), upbibtex(1), mendex(1)
.SH AUTHOR
This manual page was written by Japanese \*(TX Development Community \
<https://texjp.org>. For more information, see GitHub repository \
<https://github.com/texjporg/tex-jp-build>.


ChangeLog
2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the
	result of make install is not in itself usable.

2016-06-30  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2016 => 2017/dev.
	* tardate.ac (tex_live_tardate): 2015-05-12 => 2016-05-23.
	(Though this is not used for any release.)

2016-03-16  Karl Berry  <karl@tug.org>

	* Build: check for exitstatus.txt not having an exit status.

2016-03-06  Karl Berry  <karl@tug.org>

	* version.ac: remove /dev to prepare for 2016.

2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2015 => 2016/dev.
	* tardate.ac: Switch to 2015-06-12.

2015-05-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris (removed): Moved from here to doc/.
	* Makefile.am: Adapted.

2015-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris: New file from Mojca (without very long lines).
	* Makefile.am: Add the new file.

2015-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2015-03-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2015-03-07.

2014-11-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-11-24.

2014-10-21  Karl Berry  <karl@tug.org>

	* configure.ac: report CC/CXX values, with --version output
	if available.
	* Build: no need for objc any more.

2014-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-08-22.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2014 => 2015/dev.

texk/ptexenc/ptekf.c
/*

Copyright 2024 Japanese TeX Development Community <issue@texjp.org>
Distributed under the 3-Clause BSD License.

*/

#ifndef _WIN32
#include <unistd.h>
#endif
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include <kpathsea/getopt.h>
#include <kpathsea/progname.h>
#include <kpathsea/lib.h>
#ifdef _WIN32
#include <kpathsea/knj.h>
#ifndef CP_UTF8
#define CP_UTF8 65001
#endif
#endif
#include <ptexenc/ptexenc.h>
#include <c-auto.h>

static char opt_input_enc[5];
static char opt_output_enc[5];
static int flg_guess_enc;
static int flg_output_buffer;

#define GUESS_ONLY        -1
#define GUESS_AND_OUTPUT  +1

#define BUFFERLEN 4096
static char buff[BUFFERLEN];

static const char *optstr = "hvgbjseuGJSEU";

static struct option long_options[] = {
  {"help",    0, 0, 'h'},
  {"version", 0, 0, 'v'},
  {"guess",   0, 0, 'g'},
  {"buffer",  0, 0, 'b'},
  {0, 0, 0, 0}
};

#define MY_VERSION   "20240211"
#define BUG_ADDRESS  "issue@texjp.org"

static void show_version(void)
{
  printf("ptekf  ver." MY_VERSION " (%s) (%s, %s)\n", get_enc_string(), ptexenc_version_string, TL_VERSION);
  printf("    Copyright (C) 2024 Japanese TeX Development Community\n");
}

static void show_usage(void)
{
  printf("Usage: ptekf -[OPTION] [--] in_file1 [in_file2 ...]\n");
  printf("  j/s/e/u  Specify output encoding ISO-2022-JP, Shift_JIS, EUC-JP, UTF-8\n");
  printf("  J/S/E/U  Specify input encoding ISO-2022-JP, Shift_JIS, EUC-JP, UTF-8\n");
  printf("  G        Guess the input encoding and output to stdout or files\n");
  printf("  --guess    -g  Guess the input encoding (no conversion)\n");
  printf("  --buffer   -b  Output internal buffer without code conversion\n");
  printf("  --version  -v  Print version number\n");
  printf("  --help     -h  Print this help\n");
  printf("Default input/output encoding depends on kpathsearch parameters PTEX_KANJI_ENC, guess_input_kanji_encoding\n");
  printf("\nEmail bug reports to %s.\n", BUG_ADDRESS);
}

const char  *infname;
char        *outfname;
FILE        *infp, *outfp;

#if defined(_WIN32)
#define R_MODE "rb"
#define W_MODE "wb"
#else
#define R_MODE "r"
#define W_MODE "w"
#endif

static char *mfgets(char *buff, int size, FILE *fp)
{
  int c, len;

  if ((len = input_line2(fp, (unsigned char *)buff, NULL, 0, size, &c)) == 0
      && c != '\r' && c != '\n') return NULL;
  if (c == '\n' || c == '\r') {
    if (len+1 < size) { buff[len]=(unsigned char)c; buff[len+1]='\0'; }
    else ungetc(c, fp);
  }
  if (c == EOF) return NULL;
  return buff;
}


int
main (int argc,  char **argv)
{
  int c;
  int opt_ienc=0, opt_oenc=0;
  int (*fputs__)(const char*, FILE*);

  if (argc<=1) {
    show_version();
    show_usage();
    return 0;
  }
  kpse_set_program_name(argv[0], "ptekf");
#if defined(_WIN32)
  {
    int ac;
    char **av;
    file_system_codepage = CP_UTF8;
    is_cp932_system = 0;
    if (get_command_line_args_utf8("utf-8", &ac, &av)) {
      argv = av;
      argc = ac;
    }
  }
#endif
  enable_UPTEX(true);

  while ((c = getopt_long(argc, argv, optstr, long_options, NULL)) != -1) {
    switch (c) {

    case 'G': /* guess and output to stdout or files */
      flg_guess_enc = GUESS_AND_OUTPUT;
      opt_ienc++;
      break;

    case 'g': /* guess */
      flg_guess_enc = GUESS_ONLY;
      opt_ienc++;
      break;

    case 'b': /* output internal buffer */
      flg_output_buffer = 1;
      opt_oenc++;
      break;

    case 'h': /* help */
      show_version();
      show_usage();
      return 0;

    case 'v': /* version */
      show_version();
      return 0;

    /* input encoding */
    case 'J': strcpy(opt_input_enc,  "jis")  ; opt_ienc++; break;
    case 'S': strcpy(opt_input_enc,  "sjis") ; opt_ienc++; break;
    case 'E': strcpy(opt_input_enc,  "euc")  ; opt_ienc++; break;
    case 'U': strcpy(opt_input_enc,  "utf8") ; opt_ienc++; break;

    /* output encoding */
    case 'j': strcpy(opt_output_enc, "jis")  ; opt_oenc++; break;
    case 's': strcpy(opt_output_enc, "sjis") ; opt_oenc++; break;
    case 'e': strcpy(opt_output_enc, "euc")  ; opt_oenc++; break;
    case 'u': strcpy(opt_output_enc, "utf8") ; opt_oenc++; break;
    }
  }

  if (opt_ienc>1) {
    fprintf(stderr, "Option (--guess or -g), -G, -J, -S, -E and -U are mutually exclusive\n");
    exit(16);
  }
  if (opt_oenc>1) {
    fprintf(stderr, "Option (--buffer or -b), -j, -s, -e and -u are mutually exclusive\n");
    exit(16);
  }

  fputs__ = flg_output_buffer ? fputs : fputs2; /* without/with code conversion */

  while (optind < argc) {
    char *genc, *ret;

    infname = argv[optind++];
    if ((infp = fopen(infname, R_MODE)) == 0) {
      fprintf(stderr, "ERROR: fail to open input file [%s].", infname);
      exit(32);
    }
    if (flg_guess_enc) {
      genc = ptenc_guess_enc(infp, 1, 1);
      printf("%s: %s\n", infname, genc);
      setinfileenc(infp, genc);
      free(genc);
    }

    if ((opt_ienc && flg_guess_enc!=GUESS_ONLY) || opt_oenc) {
      if (opt_oenc) {
        outfname = xmalloc(strlen(infname)+5);
        strcpy(outfname, infname);
        strcat(outfname, ".out");

        if ((outfp = fopen(outfname, W_MODE)) == 0) {
          fprintf(stderr, "ERROR: fail to open output file [%s].", outfname);
          exit(32);
        }
      }
      else
        outfp = stdout;

      if (strlen(opt_input_enc)) {
        setinfileenc(infp, opt_input_enc);
      }
      if (strlen(opt_output_enc)) {
        setfileenc(opt_output_enc);
      }
      while ((ret = mfgets(buff, BUFFERLEN, infp)) != NULL) {
        (*fputs__)(buff, outfp);
      }
      if (ret == NULL && feof(infp)) {
        c = buff[strlen(buff)-1];
        if (c != '\n' && c != '\r')
          (*fputs__)(buff, outfp);
      }
      if (fclose(infp)) {
        fprintf(stderr, "ERROR: fail to close input file [%s].", infname);
        exit(32);
      }
      if (outfp != stdout && fclose(outfp)) {
        fprintf(stderr, "ERROR: fail to close output file [%s].", outfname);
        exit(32);
      }
      if (outfname)
        free(outfname);
    }
  }

  return 0;
}
